# Assignment 

# Part 1 
For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link a Jupyter notebook containing your solutions to the following tasks:

* First, load the dataset from the **weatherinszeged** table from Thinkful's database.
* Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 
* Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one? 
* Add *visibility* as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?
* Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

# Weather model 

We will be using the weatherinszeged  from the previous assignments to build linear regression models to predict temperature.  

Then, we will exmaine the __R-squared__ and __adjusted R-squared__ to see how explantory variables perform in each model. R-squared is a statistical measure of how close the data are to the fitted regression line. It is the percentage of the target variation that is explained by a linear model. A zero percentage indicates that the model explains none of the variability of the response data around its mean. In general, the higher the R-squared, the better the model fits the data.

We will also look at the __Akaike Information Criterion (AIC)__ and __Bayesian Information Criterion (BIC)__ when deciding which model performs the best. AIC estimates the relative amount of information lost by a given model: the less information a model loses, the higher the quality of that model. Both BIC and AIC attempt to resolve the problem of overfitting by introducing a penalty term for the number of parameters in the model; the penalty term is larger in BIC than in AIC. For both AIC and BIC, the lower the value the better. Hence, we choose the model with the lowest AIC or BIC value.

## Iteration 1 

We will start by building a linear regression model where the target variable is the difference between the apparenttemperature and the temperature. Humidity and windspeed will be our explanatory variables. 

In [1]:
# Libraries 

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy.stats as stats
from sklearn.decomposition import PCA 
from sklearn import linear_model
from sqlalchemy import create_engine
import warnings
import statsmodels.api as sm


# Import data

warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

weatherinszeged_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [2]:
# Y is the target variable
Y = weatherinszeged_df['apparenttemperature'] - weatherinszeged_df['temperature']
# X is the feature set
X = weatherinszeged_df[['humidity','windspeed']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 11 Aug 2019   Prob (F-statistic):               0.00
Time:                        23:14:32   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

$$ difference\_of\_apparenttemperature\_and\_temperature = 2.4381 - 3.0292\_humidity - 0.1193\_windspeed $$

In the tables above, we see that for our first model, R-squared is 0.288 and adjusted R-squared is 0.288. This model explains approximately 29% of the variance in the target, which is relatively low. About 71% of the target variance is unexplained by the model, so this would not be satisficatory if we produce predictions that are reasonably precise.

However, a low R-squared is not inherently bad since predicting temperature can be difficult. Since the coefficents are statistically significant, we are able to represent the mean change in the target variable for one unit of change in the explanatory variables. This still allows up to draw important conclusions on changes in the explanatory variables effect changes in the target variable. 
 

## Iteration 2 


Let's add the interaction between humidity and windspeed to see if it will improve the model.

In [3]:
weatherinszeged_df['humidity_windspeed_interaction'] = weatherinszeged_df.humidity * weatherinszeged_df.windspeed

# Y is the target variable
Y = weatherinszeged_df['apparenttemperature'] - weatherinszeged_df['temperature']
# X is the feature set
X = weatherinszeged_df[['humidity','windspeed', 'humidity_windspeed_interaction']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 11 Aug 2019   Prob (F-statistic):               0.00
Time:                        23:23:43   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================================
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              0.0839      0.033      2.511      0.012       0.018       0.149
humidity                           0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                          0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed_interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

After adding the interaction between humidity and windspeed, the R-squared and adjusted R-squared for the second model 0.341. This model captures 5.3% more of the variance explained in the target variable than the first model. 
 

## Iteration 3 

Now, we will add visibility as an explanatory variable to the first model to see if there is an improvement.

In [4]:
# Y is the target variable
Y = weatherinszeged_df['apparenttemperature'] - weatherinszeged_df['temperature']
# X is the feature set
X = weatherinszeged_df[['humidity','windspeed','visibility']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sun, 11 Aug 2019   Prob (F-statistic):               0.00
Time:                        23:33:27   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

From the summary, we can see that visibility did increase R-squared and adjusted R-squared from 0.288 to 0.304. This model is able to capture 1.6% more of the variance in the target variable than in the first model. The second model with the interaction between humidity and windspeed is able to capture the most variation in the target variable with a R-squared of 34.1%. 



### Which weather model performs best?

Overall, the second model with the addition of the interaction between humidity and wind speed outperforms the other models. In terms of R-squared, the second model is able to explain 34.1% of the variation in the target variable. While the first and third is only able to explain 28.8% and 30.4% of the variation in the target variable respectively. Looking at AIC and BIC, the second model also has the smallest criterion value of 3.388e+05, indicating that this model loses less information than the other models when adding new features.

# Part 2

In this exercise, we will work on our house prices model. We will complete the following tasks: 

* Load the **houseprices** data from Thinkful's database.
* Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.
* Do you think your model is satisfactory? If so, why?
* In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. 
* For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

# House prices model
We will revisit our house price model and try to improve performance by introducing new variables. Assessing the goodness of the fit for the model will be based on the following: 
- __R-squared__, __adjusted R-squared__: percentage of the target variation that is explained by a linear model
- __AIC__ and __BIC__: both take into consideration the sum of the squared errors (SSE), the sample size, and the number of parameters
- __F statistics__: compare two models if one of them is nested within the other

## Iteration 1

We will start by reloading our linear regression model where the target variable is sale price and the following are explanatory variables: overall quality, above grade (ground) living area, garage size, basement area, and first floor area. 

In [5]:
# Libraries 

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy.stats as stats
from sklearn.decomposition import PCA 
from sklearn import linear_model
from sqlalchemy import create_engine
import warnings
import statsmodels.api as sm


# Import data

warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

houseprices_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [6]:
# Y is the target variable
Y = houseprices_df['saleprice']
# X is the feature set
X = houseprices_df[['overallqual','grlivarea','garagearea','totalbsmtsf','firstflrsf']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.758
Model:                            OLS   Adj. R-squared:                  0.757
Method:                 Least Squares   F-statistic:                     911.5
Date:                Mon, 12 Aug 2019   Prob (F-statistic):               0.00
Time:                        00:22:33   Log-Likelihood:                -17508.
No. Observations:                1460   AIC:                         3.503e+04
Df Residuals:                    1454   BIC:                         3.506e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.034e+05   4938.326    -20.948      0.000   -1.13e+05   -9.38e+04
overallqual  2.532e+04   1055.957     23.976      0.000    2.32e+04    2.74e+04
grlivarea      43.3833      2.699     16.074      0.000      38.089      48.678
garagearea     56.6798      6.126      9.252      0.000      44.662      68.697
totalbsmtsf    22.9518      4.340      5.289      0.000      14.439      31.465
firstflrsf     11.2909      5.070      2.227      0.026       1.345      21.237
==============================================================================
Omnibus:                      507.298   Durbin-Watson:                   1.980
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            52609.779
Skew:                          -0.602   Prob(JB):                         0.00
Kurtosis:                      32.383   Cond. No.                     1.14e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.14e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In the tables above, we see that for our first model, R-squared is 0.758, adjusted R-squared is 0.757, F statistic is 911.5, AIC is 3.503 and BIC is 3.506. This is a pretty good start since our current model is able to explain about 76% of the variation in the target variable. 

Let's modify the explanatory variables to see if we can increase the R-squared and F statistis and decrease the AIC and BIC values. 

## Iteration 2 

In this iteration, we will be introducing the following categorical variables to the model: general zoning classification (mszoning) and house style.  

In [9]:
houseprices_df = pd.concat([houseprices_df,pd.get_dummies(houseprices_df.mszoning, prefix="mszoning", drop_first=True)], axis=1)
houseprices_df = pd.concat([houseprices_df,pd.get_dummies(houseprices_df.housestyle, prefix="style", drop_first=True)], axis=1)


dummy_column_names = list(pd.get_dummies(houseprices_df.mszoning, prefix="mszoning", drop_first=True).columns)
dummy_column_names = dummy_column_names + list(pd.get_dummies(houseprices_df.housestyle, prefix="style", drop_first=True).columns)

In [10]:
# Y is the target variable
Y = houseprices_df['saleprice']
# X is the feature set
X = houseprices_df[['overallqual','grlivarea','garagearea','totalbsmtsf','firstflrsf'] + dummy_column_names]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.773
Model:                            OLS   Adj. R-squared:                  0.771
Method:                 Least Squares   F-statistic:                     307.2
Date:                Mon, 12 Aug 2019   Prob (F-statistic):               0.00
Time:                        01:40:11   Log-Likelihood:                -17461.
No. Observations:                1460   AIC:                         3.496e+04
Df Residuals:                    1443   BIC:                         3.505e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.356e+05   1.28e+04    -10.590      0.000   -1.61e+05    -1.1e+05
overallqual   2.498e+04   1072.392     23.296      0.000    2.29e+04    2.71e+04
grlivarea       76.6120      6.133     12.492      0.000      64.581      88.643
garagearea      49.2236      6.119      8.045      0.000      37.221      61.226
totalbsmtsf     21.9928      4.294      5.122      0.000      13.570      30.416
firstflrsf     -28.7185      8.392     -3.422      0.001     -45.180     -12.257
mszoning_FV   1.075e+04   6673.411      1.612      0.107   -2336.393    2.38e+04
mszoning_FV   1.075e+04   6673.411      1.612      0.107   -2336.393    2.38e+04
mszoning_RH   2695.3749   7794.253      0.346      0.730   -1.26e+04     1.8e+04
mszoning_RH   2695.3749   7794.253      0.346      0.730   -1.26e+04     1.8e+04
mszoning_RL   1.157e+04   6184.775      1.870      0.062    -565.711    2.37e+04
mszoning_RL   1.157e+04   6184.775      1.870      0.062    -565.711    2.37e+04
mszoning_RM   3824.6120   6241.136      0.613      0.540   -8418.059    1.61e+04
mszoning_RM   3824.6120   6241.136      0.613      0.540   -8418.059    1.61e+04
style_1.5Unf  9961.5559   5587.902      1.783      0.075    -999.724    2.09e+04
style_1.5Unf  9961.5559   5587.902      1.783      0.075    -999.724    2.09e+04
style_1Story  1.436e+04   2563.596      5.603      0.000    9336.073    1.94e+04
style_1Story  1.436e+04   2563.596      5.603      0.000    9336.073    1.94e+04
style_2.5Fin -2.141e+04   7475.493     -2.864      0.004   -3.61e+04   -6742.693
style_2.5Fin -2.141e+04   7475.493     -2.864      0.004   -3.61e+04   -6742.693
style_2.5Unf -1.702e+04   6092.429     -2.794      0.005    -2.9e+04   -5070.499
style_2.5Unf -1.702e+04   6092.429     -2.794      0.005    -2.9e+04   -5070.499
style_2Story -1791.0257   2116.391     -0.846      0.398   -5942.559    2360.507
style_2Story -1791.0257   2116.391     -0.846      0.398   -5942.559    2360.507
style_SFoyer  1.555e+04   3860.439      4.027      0.000    7972.798    2.31e+04
style_SFoyer  1.555e+04   3860.439      4.027      0.000    7972.798    2.31e+04
style_SLvl    8689.5631   3161.476      2.749      0.006    2487.982    1.49e+04
style_SLvl    8689.5631   3161.476      2.749      0.006    2487.982    1.49e+04
==============================================================================
Omnibus:                      464.513   Durbin-Watson:                   2.005
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            46025.716
Skew:                          -0.440   Prob(JB):                         0.00
Kurtosis:                      30.492   Cond. No.                     1.05e+16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specif

In the tables above, we see that for our second model, R-squared is 0.773, adjusted R-squared is 0.771, F statistic is 309.2, AIC is 3.496 and BIC is 3.505. Compared to our first model, introducing dummy variables increased R-squared by 2% and lowered AIC and BIC by 0.1. However, the F statistic dropped by 604 points. 

## Iteration 3 

To increase the fit of our model, we will perform log transformation on sale price since it is highly skewed to the right. We'll also introduce a new variable by combining total basement area, first, and second floor area as total square feet. Lastly, we will add the interaction between total square feet and overall quality of the house. 

In [13]:
houseprices_df['totalsf'] = houseprices_df['totalbsmtsf'] + houseprices_df['firstflrsf'] + houseprices_df['secondflrsf']

houseprices_df['int_over_sf'] = houseprices_df['totalsf'] * houseprices_df['overallqual']

# Y is the target variable
Y = np.log1p(houseprices_df['saleprice'])
# X is the feature set
X = houseprices_df[['overallqual','grlivarea','garagearea','totalbsmtsf','firstflrsf', 'int_over_sf'] + dummy_column_names]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.829
Model:                            OLS   Adj. R-squared:                  0.827
Method:                 Least Squares   F-statistic:                     411.5
Date:                Mon, 12 Aug 2019   Prob (F-statistic):               0.00
Time:                        01:56:28   Log-Likelihood:                 558.37
No. Observations:                1460   AIC:                            -1081.
Df Residuals:                    1442   BIC:                            -985.6
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            9.7282      0.071    136.212      0.000       9.588       9.868
overallqual      0.1878      0.008     22.804      0.000       0.172       0.204
grlivarea        0.0004   3.36e-05     13.326      0.000       0.000       0.001
garagearea       0.0003   2.67e-05     11.761      0.000       0.000       0.000
totalbsmtsf      0.0003   2.63e-05     10.270      0.000       0.000       0.000
firstflrsf   -7.473e-05   3.66e-05     -2.040      0.041      -0.000   -2.89e-06
int_over_sf  -2.343e-05   2.71e-06     -8.636      0.000   -2.88e-05   -1.81e-05
mszoning_FV      0.2435      0.029      8.325      0.000       0.186       0.301
mszoning_FV      0.2435      0.029      8.325      0.000       0.186       0.301
mszoning_RH      0.1820      0.034      5.336      0.000       0.115       0.249
mszoning_RH      0.1820      0.034      5.336      0.000       0.115       0.249
mszoning_RL      0.2407      0.027      8.886      0.000       0.188       0.294
mszoning_RL      0.2407      0.027      8.886      0.000       0.188       0.294
mszoning_RM      0.1673      0.027      6.132      0.000       0.114       0.221
mszoning_RM      0.1673      0.027      6.132      0.000       0.114       0.221
style_1.5Unf     0.0072      0.024      0.293      0.770      -0.041       0.055
style_1.5Unf     0.0072      0.024      0.293      0.770      -0.041       0.055
style_1Story     0.0448      0.011      3.980      0.000       0.023       0.067
style_1Story     0.0448      0.011      3.980      0.000       0.023       0.067
style_2.5Fin    -0.0970      0.033     -2.945      0.003      -0.162      -0.032
style_2.5Fin    -0.0970      0.033     -2.945      0.003      -0.162      -0.032
style_2.5Unf    -0.0349      0.027     -1.312      0.190      -0.087       0.017
style_2.5Unf    -0.0349      0.027     -1.312      0.190      -0.087       0.017
style_2Story     0.0059      0.009      0.638      0.523      -0.012       0.024
style_2Story     0.0059      0.009      0.638      0.523      -0.012       0.024
style_SFoyer     0.0572      0.017      3.376      0.001       0.024       0.090
style_SFoyer     0.0572      0.017      3.376      0.001       0.024       0.090
style_SLvl       0.0430      0.014      3.111      0.002       0.016       0.070
style_SLvl       0.0430      0.014      3.111      0.002       0.016       0.070
==============================================================================
Omnibus:                      512.944   Durbin-Watson:                   2.016
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6023.099
Skew:                          -1.291   Prob(JB):                         0.00
Kurtosis:                      12.609   Cond. No.                     1.13e+16
==============================================================================

Warnings:
[1] St

In the tables above, we see that for our third model, R-squared is 0.829, adjusted R-squared is 0.827, F statistic is 411.5, AIC is -1081 and BIC is -985.6. Compared to our previous models, this model captures the most variation in the target variable. In terms of F statistics, the score is 500 lower than the first model. This model has a better AIC and BIC score than the other models. 

### Which house prices model performs best?

The third model is probably the better model in predicting sale prices. Introducing log transformation and new variables with interactions has increased the model's ability in explaining variation in the target variable. In terms of R-squared, the third model is able to explain 82.9% of the variation in the target variable. While the first and third is only able to explain 75.8% and 77.3% of the variation in the target variable respectively. Although the third model's F statistics is less than the firset model, it still outperforms the second model by 104 points. Looking at AIC and BIC, the second model also has the smallest criterion value of -1081 and -985.6, indicating that this model loses less information than the other models when adding new features.